### PR TITLE
core-initrd: move core-initrd to the snapd repo

### PR DESCRIPTION
**Pull request by alfonsosanchezbeato**
This moves https://github.com/canonical/core-initrd for 24.04 and later releases to the snapd repo. See added README.md for details.

The next step will be to build and use the resulting ubuntu-core-initramfs package to build the initramfs used in the spread tests.

### CODE REVIEW

**Comment by alfonsosanchezbeato** [comment]
The first commit is the one importing the files, most of the review should be about the rest.

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 1]
```diff
@@ -758,8 +782,9 @@ def create_initrd(parser, args):
             )
         check_call(["depmod", "-a", "-b", main, args.kernelver])
 
-        if not verify_missing_dlopen(main, os.path.join("/usr/lib", deb_arch)):
-            sys.exit(1)
+        if ubuntu_release[0] >= 25 or (ubuntu_release[0] >= 24 and ubuntu_release[1] >= 10):
```
Comment (pedronis) [inline core-initrd/latest/bin/ubuntu-core-initramfs:779]: should this check be a helper given we repeat it twice?

File: core-initrd/build-source-pkgs.sh
Comment (pedronis) [inline core-initrd/build-source-pkgs.sh:1]: this file probabbly merits a comment about what it is  useful for

**Review by pedronis — COMMENTED** [review]
did a pass, commit by commit, some of my comments might need a bit of parsing because they might apply to code that changed in later commits

File: core-initrd/latest/bin/ubuntu-core-initramfs (hunk 1 above)
Comment (alfonsosanchezbeato) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: I've created an ancillary function for this. Note though that checks might be in the future for different Ubuntu releases (we might need to do some things only from 25.04 onwards, etc.)

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 2]
```diff
@@ -735,12 +735,6 @@ def create_initrd(parser, args):
         install_systemd_files(main, rootfs, ubuntu_release)
         # Other miscelanea stuff
         install_misc(main, rootfs, deb_arch)
-        # Copy snapd bits
-        snapd_lib = path_join_make_rel_paths(rootfs, "/usr/lib/snapd")
-        snapd_files = [os.path.join(snapd_lib, "snap-bootstrap"),
-                       os.path.join(snapd_lib, "info"),
-                       "/lib/systemd/system/snapd.recovery-chooser-trigger.service"]
-        install_files(snapd_files, main, rootfs)
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: Yes, that is correct

**Review by alfonsosanchezbeato — COMMENTED** [review]
@pedronis thanks, comments addressed now

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 3]
```diff
@@ -315,10 +315,13 @@ def package_files(pkgs, sysroot):
...
 
 
-def install_systemd_files(dest_dir, sysroot):
+def install_systemd_files(dest_dir, sysroot, ubuntu_release):
     # Build list of files and directories
 
-    lines = package_files(["systemd", "systemd-sysv", "systemd-cryptsetup"], sysroot)
+    syd_pkgs = ["systemd", "systemd-sysv"]
+    if ubuntu_release[0] >= 25 or (ubuntu_release[0] >= 24 and ubuntu_release[1] >= 10):
+        syd_pkgs.append("systemd-cryptsetup")
```
[the remark this answers is not in the document]
Comment (pedronis) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: thx, I noticed only later that we have also 24.04 support

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 4]
```diff
@@ -319,7 +319,7 @@ def install_systemd_files(dest_dir, sysroot, ubuntu_release):
     # Build list of files and directories
 
     syd_pkgs = ["systemd", "systemd-sysv"]
-    if ubuntu_release[0] >= 25 or (ubuntu_release[0] >= 24 and ubuntu_release[1] >= 10):
+    if release_higher_or_equal(ubuntu_release, [24, 10]):
```
Comment (pedronis) [inline core-initrd/latest/bin/ubuntu-core-initramfs:322]: thx, I would use a tuple there (24, 10)

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 5]
```diff
@@ -776,7 +783,7 @@ def create_initrd(parser, args):
             )
         check_call(["depmod", "-a", "-b", main, args.kernelver])
 
-        if ubuntu_release[0] >= 25 or (ubuntu_release[0] >= 24 and ubuntu_release[1] >= 10):
+        if release_higher_or_equal(ubuntu_release, [24, 10]):
```
Comment (pedronis) [inline core-initrd/latest/bin/ubuntu-core-initramfs:786]: same

**Review by pedronis — APPROVED** [review]
thx, small idiomatic python nitpick

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 6]
```diff
@@ -668,13 +672,32 @@ def verify_missing_dlopen(destdir, libdir):
...
                         "DEB_HOST_MULTIARCH"], env=proc_env).decode("utf-8")
     return out.splitlines()[0]
 
+
+def get_ubuntu_release(sysroot):
+    with open(os.path.join(sysroot, "etc/os-release"), "r") as f:
+        data = f.read()
+        regex = re.compile('^VERSION_ID="(.*)"$')
+        for line in data.splitlines():
+            res = regex.search(line)
```
Comment (valentindavid) [inline core-initrd/latest/bin/ubuntu-core-initramfs:689]: Do we need to do it for every line? Cannot we search in the whole data?

File: core-initrd/latest/debian/rules [hunk 7]
```diff
@@ -1,16 +1,60 @@
...
+
+# check if we need to include the testkeys in the binary
+ifneq (,$(filter testkeys,$(DEB_BUILD_OPTIONS)))
+# if enabled also enable bootloader assets testing and fault injection
+_SNAP_TAGS:=nomanagers,withtestkeys,faultinject
+else
+_SNAP_TAGS:=nomanagers
+endif
+
+ifeq (${FIPSBUILD},1)
```
Comment (valentindavid) [inline core-initrd/latest/debian/rules:26]: Do we have fips version of pc-kernel?

File: core-initrd/latest/debian/rules [hunk 8]
```diff
@@ -54,6 +54,8 @@ override_dh_auto_install:
 # snapd bits
 	mkdir -p debian/tmp/usr/lib/snapd/
 	cp snap-bootstrap debian/tmp/usr/lib/snapd/
+	cp snapd/info debian/tmp/usr/lib/snapd/
+	cp snapd/snapd.recovery-chooser-trigger.service debian/tmp/usr/lib/systemd/system/
```
Comment (valentindavid) [inline core-initrd/latest/debian/rules:58]: Cannot we use a `.install` for this?

File: core-initrd/latest/debian/rules (hunk 7 above)
Comment (Meulengracht) [reply core-initrd/latest/debian/rules:26]: there exists one, but i dunno if it's published

File: core-initrd/latest/debian/rules (hunk 8 above)
Comment (alfonsosanchezbeato) [reply core-initrd/latest/debian/rules]: Done

File: core-initrd/latest/bin/ubuntu-core-initramfs (hunk 6 above)
Comment (alfonsosanchezbeato) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: Done

File: core-initrd/latest/bin/ubuntu-core-initramfs [hunk 9]
```diff
@@ -668,13 +672,32 @@ def verify_missing_dlopen(destdir, libdir):
...
     proc_env["DPKG_DATADIR"] = sysroot + "/usr/share/dpkg"
     out = check_output(["dpkg-architecture", "-q",
                         "DEB_HOST_MULTIARCH"], env=proc_env).decode("utf-8")
     return out.splitlines()[0]
 
+
+def get_ubuntu_release(sysroot):
+    with open(os.path.join(sysroot, "etc/os-release"), "r") as f:
+        data = f.read()
+        regex = re.compile('^VERSION_ID="(.*)"$')
```
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: Done, but not using the placeholder and with re.MULTILINE

File: core-initrd/latest/bin/ubuntu-core-initramfs (hunk 5 above)
Comment (alfonsosanchezbeato) [reply core-initrd/latest/bin/ubuntu-core-initramfs]: Done

File: core-initrd/latest/debian/rules (hunk 7 above)
Comment (alfonsosanchezbeato) [reply core-initrd/latest/debian/rules:26]: We will want it at some point

**Comment by alfonsosanchezbeato** [comment]
@Meulengracht @valentindavid thanks for the reviews, I've addressed your comments now

**Review by Meulengracht — APPROVED** [review]
lgtm

**Comment by alfonsosanchezbeato** [comment]
Cleaned up PR commit history, as it is worth to do a rebase and merge in this case to distinguish between the import and the other changes.

**Pull request merged**